mllib/src/main/scala/org/apache/spark/ml/stat/ANOVATest.scala

```diff
@@ -80,65 +81,141 @@ object ANOVATest {
+      (0.0, 0.0, new OpenHashMap[Double, Double]))(
+      seqOp = {
+        case ((sum, sumOfSq, sums), (label, value)) =>
           // sums: mapOfSumPerClass (key: label, value: sum of features for each label)
-          //                                         ( 1 -> 3.3 + 1.0, 2 -> 2.5, 3 -> 3.0 + 2.0 )
-          // counts: mapOfCountPerClass (key: label, value: count of features for each label)
-          //                                         ( 1 -> 2, 2 -> 2, 3 -> 2 )
-          // sqSum: square of sum of all data ((3.3+2.5+1.0+3.0+2.0)^2)
-          val sqSum = sum * sum
-          val ssTot = sumOfSq - sqSum / numSamples
-
-          // sumOfSqSumPerClass:
-          //     sum( sq_sum_classes[k] / n_samples_per_class[k] for k in range(n_classes))
-          //     e.g. ((3.3+1.0)^2 / 2 + 2.5^2 / 1 + (3.0+2.0)^2 / 2)
-          val sumOfSqSumPerClass = sums.iterator
-            .map { case (label, sum) => sum * sum / counts(label) }.sum
-          // Sums of Squares Between
-          val ssbn = sumOfSqSumPerClass - (sqSum / numSamples)
-          // Sums of Squares Within
-          val sswn = ssTot - ssbn
-          // degrees of freedom between
-          val dfbn = numClasses - 1
-          // degrees of freedom within
-          val dfwn = numSamples - numClasses
-          // mean square between
-          val msb = ssbn / dfbn
-          // mean square within
-          val msw = sswn / dfwn
-          val fValue = msb / msw
-          val pValue = 1 - new FDistribution(dfbn, dfwn).cumulativeProbability(fValue)
-          (col, pValue, dfbn + dfwn, fValue)
-        }.collect().sortBy(_._1).map {
-          case (col, pValue, degreesOfFreedom, fValue) =>
-            new ANOVATestResult(pValue, degreesOfFreedom, fValue)
-        }
+          sums.changeValue(label, value, _ + value)
+          (sum + value, sumOfSq + value * value, sums)
+      },
+      combOp = {
+        case ((sum1, sumOfSq1, sums1), (sum2, sumOfSq2, sums2)) =>
+          sums2.foreach { case (v, w) => sums1.changeValue(v, w, _ + w) }
+          (sum1 + sum2, sumOfSq1 + sumOfSq2, sums1)
+      }
+    ).mapValues { case (sum, sumOfSq, sums) =>
+      val counts = bcCounts.value
+      counts.keysIterator.foreach { label =>
+        // adjust sums if all related feature values are 0 for some label
+        if (!sums.contains(label)) sums.update(label, 0.0)
+      }
+      computeANOVA(sum, sumOfSq, sums.toMap, counts)
+    }.collectAsMap()
+
+    bcCounts.destroy()
+
+    val finalResults = Array.ofDim[SelectionTestResult](numFeatures)
+    results.foreach { case (col, (pValue, degreesOfFreedom, fValue)) =>
+      finalResults(col) = new ANOVATestResult(pValue, degreesOfFreedom, fValue)
+    }
+
+    if (results.size < numFeatures) {
+      // if some column only contains 0 values
+      val (pValue, degreesOfFreedom, fValue) =
+        computeANOVA(0.0, 0.0, counts.mapValues(_ => 0.0), counts)
```

Comment:
For the category that only contains value 0s, the ```pValue``` and ```fValue``` are NaN. Maybe just get the DF and skip calling ```computeANOVA```?

Reply:
Yes, because the result contains `degreesOfFreedom` which is related to the input `counts`, I prefer to call `computeANOVA` here.

Reply:
```degreesOfFreedom = numSamples - 1```, right? no need to call ```computeANOVA```?

Reply:
yes, but I think it maybe more consistent to return a result via `computeANOVA`